examples/gitops/argocd-multi-cluster/hub-cluster/main.tf

```diff
@@ -0,0 +1,372 @@
+    clientID     = var.argocd_sso_client_id
+    clientSecret = var.argocd_sso_client_secret
+    logoutURL    = "${var.argocd_sso_logout_url}?client_id=${var.argocd_sso_client_id}&logout_uri=https://${local.argocd_subdomain}.${local.domain_name}/logout"
+    cliClientID  = var.argocd_sso_cli_client_id
+    url          = "https://${local.argocd_subdomain}.${local.domain_name}"
+  }) : ""
+
+}
+
+################################################################################
+# Cluster
+################################################################################
+
+#tfsec:ignore:aws-eks-enable-control-plane-logging
+module "eks" {
+  source  = "terraform-aws-modules/eks/aws"
+  version = "~> 19.10"
+
+  cluster_name                   = local.name
+  cluster_version                = local.cluster_version
+  cluster_endpoint_public_access = true
+
+  vpc_id     = module.vpc.vpc_id
+  subnet_ids = module.vpc.private_subnets
+
+  eks_managed_node_groups = {
+    initial = {
+      instance_types = [local.instance_type]
+
+      min_size     = 3
+      max_size     = 10
+      desired_size = 5
+    }
+  }
+
+  tags = local.tags
+}
+
+################################################################################
+# Kubernetes Addons
+################################################################################
+
+module "eks_blueprints_kubernetes_addons" {
+  source = "github.com/csantanapr/terraform-aws-eks-blueprints-addons?ref=argo-multi-cluster" #TODO change git org to aws-ia
+
+  eks_cluster_id       = module.eks.cluster_name
+  eks_cluster_endpoint = module.eks.cluster_endpoint
+  eks_oidc_provider    = module.eks.oidc_provider
+  eks_cluster_version  = module.eks.cluster_version
+
+  enable_argocd         = true
+  argocd_manage_add_ons = true # Indicates addons to be install via ArgoCD
+  argocd_helm_config = {
+    namespace = local.argocd_namespace
+    version   = "5.23.1" # ArgoCD v2.6.2
+    values    = [local.argocd_values, local.argocd_sso]
+  }
+
+  # Add-ons
+  enable_aws_load_balancer_controller = true                                # ArgoCD UI depends on aws-loadbalancer-controller for Ingress
+  enable_metrics_server               = true                                # ArgoCD HPAs depend on metric-server
+  enable_external_dns                 = local.enable_ingress ? true : false # ArgoCD Server and UI use valid https domain name
+  external_dns_helm_config = {
+    domainFilters : [local.domain_name]
+  }
+  external_dns_route53_zone_arns = [local.argocd_domain_arn] # ArgoCD Server and UI domain name is registered in Route 53
+
+  # Observability for ArgoCD
+  enable_amazon_eks_aws_ebs_csi_driver = true
+  enable_prometheus                    = true
+
+  tags = local.tags
+}
+
+################################################################################
+# EKS Blueprints Add-Ons via ArgoCD
+################################################################################
+
+module "eks_blueprints_argocd_addons" {
```

Comment:
what is the difference between the module block above and this one? (meaning - what is the motivation for two separate definitions?)

Reply:
or rather - theres several used here - just curious why they need to be split out

Reply:
Is to allow a clean destroy of dependencies (open to better ideas)
On deploy:
1. eks_blueprints_kubernetes_addons (this one installs argocd controller)
2. eks_blueprints_argocd_addons (this one deploys addons, like aws-load-balancer)
3. eks_blueprints_argocd_workloads (this one deploys the apps which depend on addons like aws-load-balancer)
On destroy
1. eks_blueprints_argocd_workloads (first workloads, before addons (ie aws-load-balancer) and argocd controller)
2. eks_blueprints_argocd_addons (after workloads are removed, before addons (ie aws-load-balancer) and argocd controller)
3. eks_blueprints_kubernetes_addons (after worloads and addons, to remove the argo controller)